README.md

```diff
@@ -539,30 +524,271 @@ services.Configure<TargetingEvaluationOptions>(options =>
+There are scenarios which require the state of a feature to remain consistent during the lifetime of a request. The values returned from the standard `IFeatureManager` may change if the `IConfiguration` source which it is pulling from is updated during the request. This can be prevented by using `IFeatureManagerSnapshot`. `IFeatureManagerSnapshot` can be retrieved in the same manner as `IFeatureManager`. `IFeatureManagerSnapshot` implements the interface of `IFeatureManager`, but it caches the first evaluated state of a feature during a request and will return the same state of a feature during its lifetime. Symmetric functionality is available for dynamic features through the use of `IDynamicFeatureManagerSnapshot`.
 
 ## Custom Feature Providers
 
 Implementing a custom feature provider enable developers to pull feature flags from sources such as a database or a feature management service. The included feature provider that is used by default pulls feature flags from .NET Core's configuration system. This allows for features to be defined in an [appsettings.json](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.1#jcp) file or in configuration providers like [Azure App Configuration](https://docs.microsoft.com/en-us/azure/azure-app-configuration/quickstart-feature-flag-aspnet-core?tabs=core2x). This behavior can be substituted to provide complete control of where feature definitions are read from.
 
-To customize the loading of feature definitions, one must implement the `IFeatureDefinitionProvider` interface.
+To customize the loading of feature definitions, one must implement the `IFeatureFlagDefinitionProvider` interface.
 
-```
-public interface IFeatureDefinitionProvider
+``` C#
+public interface IFeatureFlagDefinitionProvider
 {
-    Task<FeatureDefinition> GetFeatureDefinitionAsync(string featureName);
+        Task<FeatureFlagDefinition> GetFeatureFlagDefinitionAsync(string featureName, CancellationToken cancellationToken = default);
 
-    IAsyncEnumerable<FeatureDefinition> GetAllFeatureDefinitionsAsync();
+        IAsyncEnumerable<FeatureFlagDefinition> GetAllFeatureFlagDefinitionsAsync(CancellationToken cancellationToken = default);
 }
 ```
 
 To use an implementation of `IFeatureDefinitionProvider` it must be added into the service collection before adding feature management. The following example adds an implementation of `IFeatureDefinitionProvider` named `InMemoryFeatureDefinitionProvider`.
 
-```
+``` C#
 services.AddSingleton<IFeatureDefinitionProvider, InMemoryFeatureDefinitionProvider>()
         .AddFeatureManagement()
 ```
 
+It is also possible to provide custom dynamic feature definitions. This is done by implementing the `IDynamicFeatureDefinitionProvider` interface and registering it as mentioned above.
+
+## Dynamic Features

```

Comment:
Many sections like Caching and Custom Feature Providers apply to both feature flags and dynamic features, should we move them _after_ the Dynamic Features section?

Reply:
Good idea. Done.